### ISSUE TITLE
LayerNorm and RMSNorm size the per-program weight-gradient buffer by floor division while launching a ceiling-sized grid, so the last block is written out of bounds and dropped from the gradient

### ISSUE THREAD

**Issue by truong-v** [BobMcDear/attorch#18]
## Summary

Both norm layers allocate the per-program weight-gradient buffer with `batch_dim // BLOCK_SIZE_BATCH` but launch `cdiv(batch_dim, BLOCK_SIZE_BATCH)` programs. Whenever the block size does not divide the batch dimension there is exactly one program more than the buffer has rows, and the kernel indexes that buffer by the raw program id with no bound check — only the feature axis is masked. So the last program stores a full feature row past the end of the tensor: an out-of-bounds **write**, which a caching allocator normally hides, and — because that row lands outside the tensor — a row missing from the `weight_grad.sum(dim=0)` that follows, so the weight gradient silently loses the last block of rows. LayerNorm allocates `bias_grad` the same way, so it overruns and truncates two tensors.

`BLOCK_SIZE_BATCH > 1` needs `feat_dim < 64` and `batch_dim >= 1024` — a large batch of narrow rows. `tests/utils.py::default_shapes` does contain three shapes that reach `BLOCK_SIZE_BATCH > 1`, but all three have batch dimensions their block size divides evenly (8096/8, 12288/16, 131072/128), so the grid and the buffer agree and nothing has caught this. One odd batch dimension is enough.

## Reproduction

```python
import torch
from torch import nn

import attorch

for shape in [(2049, 2), (4097, 4)]:
    torch.manual_seed(0)
    base = torch.randn(shape, device='cuda')
    grad = torch.randn(shape, device='cuda')

    a_in = base.clone().requires_grad_(True)
    t_in = base.clone().requires_grad_(True)
    a_norm = attorch.LayerNorm(shape[-1])
    t_norm = nn.LayerNorm(shape[-1], device='cuda')
    for mod in (a_norm, t_norm):
        torch.manual_seed(0)
        nn.init.normal_(mod.weight)
        nn.init.zeros_(mod.bias)

    a_norm(a_in).backward(grad)
    t_norm(t_in).backward(grad)
    print(shape,
          'weight-grad max|diff|', (a_norm.weight.grad - t_norm.weight.grad).abs().max().item(),
          'bias-grad max|diff|', (a_norm.bias.grad - t_norm.bias.grad).abs().max().item())
```

```
(2049, 2) weight-grad max|diff| 0.5584754943847656 bias-grad max|diff| 2.404165267944336
(4097, 4) weight-grad max|diff| 2.67852783203125 bias-grad max|diff| 1.877471923828125
```

The input gradient is unaffected, which isolates this to the buffer. The write itself, under `PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer --tool memcheck` on the same `(2049, 2)` LayerNorm backward — `2049 // 2 = 1024` rows of 2 floats is 8192 bytes, and program 1024 writes just past it:

```
========= Invalid __global__ write of size 4 bytes
=========     at layer_norm_backward_kernel+0x4b0 in attorch/layer_norm_kernels.py:223
=========     by thread (0,0,0) in block (1024,0,0)
=========     Address 0x7f9596219000 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f9596217000 of size 8,192 bytes
========= ERROR SUMMARY: 98 errors
```

Batch dimensions the block size does divide, and any `feat_dim >= 64` (where `BLOCK_SIZE_BATCH` is 1), match PyTorch to float noise.

## Environment

- attorch at `88ed34d`
- NVIDIA L40S (sm_89)
- torch 2.4.0+cu121, triton 3.0.0, Python 3.10, driver CUDA 12.4
